doc/specification.md

```diff
@@ -0,0 +1,54 @@
+# Test Specification
+
+This document describes how to use the JSON test specifications to write your own test executor for
+[`wasi-testsuite`](..). The test executor included in this project provides a completely-usable
+reference [implementation](../test-runner), but projects with other requirements may want to run the
+tests in their own way (e.g., no Python dependency). The JSON test specifications provide a simple
+way to verify that the tests indeed passed.
+
+### Configuration
+
+Before executing anything, a test executor is expected to:
+- find all `*.wasm` files in a given subdirectory &mdash; these are the _test cases_
+- find all `*.cleanup` files in a given subdirectory and remove them &mdash; these are test
+  artifacts that can be generated during testing
+- for each test case, look for a `.json` file in the same directory matching the base name (e.g.,
+  `foo.json` for `foo.wasm`) &mdash; parse this _specification_
+- if no `.json` file is present, use a default specification; a (conceptual) default specification
+  would look like:
+
+  ```json
+  {
+    "args": [],
+    "dirs": [],
+    "env": {},
+    "exit_code": 0,
+    "stderr": "",
+    "stdout": ""
+  }
+  ```
+
+- if the specification is missing fields, use default values
+
+### Execution
+
+To execute the tests, the test executor is expected to:
+- `env`: construct an environment from each key-value pair in `env`; the environment should only
+  contain these keys and otherwise should be empty (note that this environment is the WASI
+  environment, whether the engine inherits the shell environment or explicitly configures it via
+  `--env` parameters)
```

Comment:
the `--env` option is runtime-specific i guess.

Reply:
Yeah, the parenthetical is giving an example.